datamodel/include/datamodel/rendering/ShaderProgram.hpp

```diff
@@ -0,0 +1,50 @@
+#ifndef RCBE_ENGINE_SHADERPROGRAM_HPP
+#define RCBE_ENGINE_SHADERPROGRAM_HPP
+
+#include <vector>
+#include <memory>
+
+#include <rcbe-engine/datamodel/math/Vector.hpp>
+#include <rcbe-engine/datamodel/math/Matrix.hpp>
+#include <rcbe-engine/datamodel/math/MatrixColumnMajorAdaptor.hpp>
+
+namespace rcbe::rendering {
+
+class Shader;
+
+using ShaderCollection = std::vector<Shader>;
+
+class ShaderProgram {
+public:
+    using ShaderProgramHandle = size_t;
+
+    ShaderProgram() = delete;
+    explicit ShaderProgram(ShaderCollection &&shaders);
+    ~ShaderProgram();
+
+    ShaderProgram(const ShaderProgram &other) = delete;
+    ShaderProgram &operator=(const ShaderProgram &other) = delete;
+
+    void use() const;
+
+    void setBool(const std::string &name, bool value) const;
+    void setInteger(const std::string &name, int value) const;
+    void setFloat(const std::string &name, float value) const;
+    void setVector(const std::string &name, const math::Vector4f &value) const;
+    void setVector(const std::string &name, const math::Vector3f &value) const;
+    void setVector(const std::string &name, const math::Vector2f &value) const;
+    void setMatrix(const std::string &name, const math::MatrixColumnMajorAdaptor<core::EngineScalar> &value) const;
+    void setMatrix(const std::string &name, const math::MatrixColumnMajorAdaptor<float> &value) const;
+    void setMatrix(const std::string &name, const math::Matrix4x4f &value) const;
```

Comment:
Just a suggestion. Implement set functions for all other needed types (i.e. Matrix3x3 etc.)

Reply:
Done.